File: 025 Reverse Nodes in k-Group/reverse_node_in_k_groups_iterative.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        if not head or not head.next or k < 2:
            return head
        
        dummy = ListNode(0)
        dummy.next = head
        curr = dummy
        tail = dummy
        
        while True:
            count = 0
            while count < k and tail:
                tail = tail.next;
                count += 1
            
            # insufficient k nodes
            if not tail:
                break

            # for next cycle 
            head = curr.next
            while count > 1:
                # delete temp
                temp = curr.next
                curr.next = temp.next
                # insert temp
                temp.next = tail.next
                tail.next = temp;
                count -= 1
            # for next loop
            curr = tail = head
        return dummy.next

File: 0025 Reverse Nodes in k-Group/ReverseNodeInKGroups_Iterative.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode curr = dummy;
        ListNode tail = dummy; 
        
        while (true) {
            int count = 0;
            // find the kth node if exist
            while (tail != null && count < k) {
                tail = tail.next;
                count++;
            }
            
            // not enough listnodes
            if (tail == null) {
                break;
            }
        
            head = curr.next;
            // reverse k - 1 times
            while (count > 1) {
                // delete temp
                ListNode temp = curr.next;
                curr.next = temp.next;
                // insert temp
                temp.next = tail.next;
                tail.next = temp;
                count--;     
            }
            curr = head;
            tail = head;
            
        }
        return dummy.next;
    }
}